cadc-util/src/main/java/ca/nrc/cadc/auth/TokenValidator.java

```diff
@@ -0,0 +1,158 @@
+     * CADC AuthenticatorImpl(s).
+     * 
+     * @param subject The subject containing extracted principals
+     * @return The same subject but with associated public credentials
+     * @throws AccessControlException 
+     */
+    public static Subject validateTokens(Subject subject) throws NotAuthenticatedException {
+        
+        // cookies
+        Set<CookiePrincipal> cookiePrincipals = subject.getPrincipals(CookiePrincipal.class);
+        log.debug("validateTokens: found " + cookiePrincipals.size() + " cookie principals");
+        if (!cookiePrincipals.isEmpty()) {
+            SSOCookieManager ssoCookieManager = new SSOCookieManager();
+            for (CookiePrincipal p : cookiePrincipals) {
+                try {
+                    SignedToken cookieToken = ssoCookieManager.parse(p.getValue());
+                    subject.getPrincipals().addAll(cookieToken.getIdentityPrincipals());
+                    List<SSOCookieCredential> cookieCredentialList =
+                        ssoCookieManager.getSSOCookieCredentials(p.getValue());
+                    log.debug("Adding " + cookieCredentialList.size() + " cookie credentials to subject");
+                    subject.getPublicCredentials().addAll(cookieCredentialList);
+                } catch (InvalidDelegationTokenException ex) {
+                    throw new NotAuthenticatedException("invalid cookie: " + ex.getMessage(), ex);
+                }
+            }
+        }
+        
+        // tokens
+        Set<AuthorizationTokenPrincipal> tokenPrincipals = subject.getPrincipals(AuthorizationTokenPrincipal.class);
+        log.debug("validateTokens: found " + tokenPrincipals.size() + " token principals");
+        for (AuthorizationTokenPrincipal p : tokenPrincipals) {
+            String token = p.getName();
+            // parse the token into challenge type and credentials.  If only credentials available (no type), assume
+            // the deprecated delegation token
+            String challengeType = null;
+            String credentials = null;
+            int spaceIndex = token.indexOf(" ");
+            if (spaceIndex == -1) {
+                challengeType = AuthenticationUtil.TOKEN_TYPE_CADC;
+                credentials = token;
+            } else {
+                challengeType = token.substring(0, spaceIndex).trim();
+                credentials = token.substring(spaceIndex + 1).trim();
+                if (!AuthenticationUtil.CHALLENGE_TYPE_BEARER.equals(challengeType)
+                    && !AuthenticationUtil.CHALLENGE_TYPE_IVOA.equals(challengeType)) {
+                    throw new NotAuthenticatedException(challengeType, OAuthError.INVALID_REQUEST,
+                        "unsupported challenge type: " + challengeType);
+                }
+            }
+            
+            try {
+                SignedToken validatedToken = SignedToken.parse(credentials);
+                subject.getPrincipals().add(validatedToken.getUser());
```

Comment:
I think you need to remove the AuthorizationTokenPrincipal from the subject. It's only a transient and no longer needed.